File: awesomestuff/src/main/java/com/awesome/dialog/AwesomeTextPromptDialog.java
package com.awesome.dialog;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.StringRes;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import com.awesome.R;


/**
 * Created by yarolegovich on 16.04.2016.
 */
public class AwesomeTextPromptDialog extends AwesomeBase<AwesomeTextPromptDialog> {

    private static final String KEY_HAS_ERROR = "key_has_error";
    private static final String KEY_TYPED_TEXT = "key_typed_text";

    private EditText inputField;
    private TextView errorMessage;
    private TextView confirmButton, negativeButton;

    private TextFilter filter;

    public AwesomeTextPromptDialog(Context context) {
        super(context);
    }

    public AwesomeTextPromptDialog(Context context, int theme) {
        super(context, theme);
    }

    {
        confirmButton = findView(R.id.ld_btn_confirm);
        negativeButton = findView(R.id.ld_btn_negative);
        inputField = findView(R.id.ld_text_input);
        errorMessage = findView(R.id.ld_error_message);
        inputField.addTextChangedListener(new HideErrorOnTextChanged());
    }

    public AwesomeTextPromptDialog setConfirmButton(@StringRes int text, OnTextInputConfirmListener listener) {
        return setConfirmButton(string(text), listener);
    }

    public AwesomeTextPromptDialog setConfirmButton(String text, OnTextInputConfirmListener listener) {
        confirmButton.setText(text);
        confirmButton.setOnClickListener(new TextInputListener(listener));
        return this;
    }

    public AwesomeTextPromptDialog setConfirmButtonColor(int color) {
        confirmButton.setTextColor(color);
        return this;
    }

    public AwesomeTextPromptDialog setNegativeButton(@StringRes int text, View.OnClickListener listener){
        return setNegativeButton(string(text), listener);
    }

    public AwesomeTextPromptDialog setNegativeButton(String text, View.OnClickListener listener){
        negativeButton.setVisibility(View.VISIBLE);
        negativeButton.setText(text);
        negativeButton.setOnClickListener(new ClickListenerDecorator(listener, true));
        return this;
    }

    public AwesomeTextPromptDialog setNegativeButtonColor(int color) {
        negativeButton.setTextColor(color);
        return this;
    }

    public AwesomeTextPromptDialog setInputFilter(@StringRes int errorMessage, TextFilter filter) {
        return setInputFilter(string(errorMessage), filter);
    }

    public AwesomeTextPromptDialog setInputFilter(String errorMessage, TextFilter filter) {
        this.filter = filter;
        this.errorMessage.setText(errorMessage);
        return this;
    }

    public AwesomeTextPromptDialog setErrorMessageColor(int color) {
        errorMessage.setTextColor(color);
        return this;
    }

    public AwesomeTextPromptDialog setInputType(int inputType) {
        inputField.setInputType(inputType);
        return this;
    }

    public AwesomeTextPromptDialog addTextWatcher(TextWatcher textWatcher) {
        inputField.addTextChangedListener(textWatcher);
        return this;
    }

    public AwesomeTextPromptDialog setInitialInput(@StringRes int text) {
        return setInitialInput(string(text));
    }

    public AwesomeTextPromptDialog setInitialInput(String text) {
        inputField.setText(text);
        return this;
    }

    public AwesomeTextPromptDialog setHint(@StringRes int hint) {
        return setHint(string(hint));
    }

    public AwesomeTextPromptDialog setHint(String text) {
        inputField.setHint(text);
        return this;
    }

    private void setError() {
        errorMessage.setVisibility(View.VISIBLE);
    }

    private void hideError() {
        errorMessage.setVisibility(View.GONE);
    }

    @Override
    void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(KEY_HAS_ERROR, errorMessage.getVisibility() == View.VISIBLE);
        outState.putString(KEY_TYPED_TEXT, inputField.getText().toString());
    }

    @Override
    void restoreState(Bundle savedState) {
        super.restoreState(savedState);
        if (savedState.getBoolean(KEY_HAS_ERROR, false)) {
            setError();
        }
        inputField.setText(savedState.getString(KEY_TYPED_TEXT));
    }

    @Override
    protected int getLayout() {
        return R.layout.dialog_text_input;
    }

    private class TextInputListener implements View.OnClickListener {

        private OnTextInputConfirmListener wrapped;

        private TextInputListener(OnTextInputConfirmListener wrapped) {
            this.wrapped = wrapped;
        }

        @Override
        public void onClick(View v) {
            String text = inputField.getText().toString();

            if (filter != null) {
                boolean isWrongInput = !filter.check(text);
                if (isWrongInput) {
                    setError();
                    return;
                }
            }

            if (wrapped != null) {
                wrapped.onTextInputConfirmed(text);
            }

            dismiss();
        }
    }

    private class HideErrorOnTextChanged implements TextWatcher {

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            hideError();
        }

        @Override
        public void afterTextChanged(Editable s) {

        }
    }

    public interface OnTextInputConfirmListener {
        void onTextInputConfirmed(String text);
    }

    public interface TextFilter {
        boolean check(String text);
    }
}

File: fusion/src/main/java/com/awesome/dialog/AwesomeTextPromptDialog.kt
package com.awesome.dialog

import android.content.Context
import android.os.Bundle
import android.support.annotation.StringRes
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.TextView

import com.awesome.R


/**
 * Created by Sandip Savaliya on 16.04.2016.
 */
class AwesomeTextPromptDialog : AwesomeBase<AwesomeTextPromptDialog> {

    private var inputField: EditText? = null
    private var errorMessage: TextView? = null
    private var confirmButton: TextView? = null
    private var negativeButton: TextView? = null

    private var filter: TextFilter? = null

    protected override val layout: Int
        get() = R.layout.dialog_text_input

    constructor(context: Context) : super(context) {}

    constructor(context: Context, theme: Int) : super(context, theme) {}

    init {
        confirmButton = findView(R.id.ld_btn_confirm)
        negativeButton = findView(R.id.ld_btn_negative)
        inputField = findView(R.id.ld_text_input)
        errorMessage = findView(R.id.ld_error_message)
        inputField!!.addTextChangedListener(HideErrorOnTextChanged())
    }

    fun setConfirmButton(@StringRes text: Int, listener: OnTextInputConfirmListener): AwesomeTextPromptDialog {
        return setConfirmButton(string(text), listener)
    }

    fun setConfirmButton(text: String, listener: OnTextInputConfirmListener): AwesomeTextPromptDialog {
        confirmButton!!.text = text
        confirmButton!!.setOnClickListener(TextInputListener(listener))
        return this
    }

    fun setConfirmButtonColor(color: Int): AwesomeTextPromptDialog {
        confirmButton!!.setTextColor(color)
        return this
    }

    fun setNegativeButton(@StringRes text: Int, listener: View.OnClickListener): AwesomeTextPromptDialog {
        return setNegativeButton(string(text), listener)
    }

    fun setNegativeButton(text: String, listener: View.OnClickListener): AwesomeTextPromptDialog {
        negativeButton!!.visibility = View.VISIBLE
        negativeButton!!.text = text
        negativeButton!!.setOnClickListener(AwesomeBase.ClickListenerDecorator(listener, true, dialog!!))
        return this
    }

    fun setNegativeButtonColor(color: Int): AwesomeTextPromptDialog {
        negativeButton!!.setTextColor(color)
        return this
    }

    fun setInputFilter(@StringRes errorMessage: Int, filter: TextFilter): AwesomeTextPromptDialog {
        return setInputFilter(string(errorMessage), filter)
    }

    fun setInputFilter(errorMessage: String, filter: TextFilter): AwesomeTextPromptDialog {
        this.filter = filter
        this.errorMessage!!.text = errorMessage
        return this
    }

    fun setErrorMessageColor(color: Int): AwesomeTextPromptDialog {
        errorMessage!!.setTextColor(color)
        return this
    }

    fun setInputType(inputType: Int): AwesomeTextPromptDialog {
        inputField!!.inputType = inputType
        return this
    }

    fun addTextWatcher(textWatcher: TextWatcher): AwesomeTextPromptDialog {
        inputField!!.addTextChangedListener(textWatcher)
        return this
    }

    fun setInitialInput(@StringRes text: Int): AwesomeTextPromptDialog {
        return setInitialInput(string(text))
    }

    fun setInitialInput(text: String): AwesomeTextPromptDialog {
        inputField!!.setText(text)
        return this
    }

    fun setHint(@StringRes hint: Int): AwesomeTextPromptDialog {
        return setHint(string(hint))
    }

    fun setHint(text: String): AwesomeTextPromptDialog {
        inputField!!.hint = text
        return this
    }

    private fun setError() {
        errorMessage!!.visibility = View.VISIBLE
    }

    private fun hideError() {
        errorMessage!!.visibility = View.GONE
    }

    internal override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(KEY_HAS_ERROR, errorMessage!!.visibility == View.VISIBLE)
        outState.putString(KEY_TYPED_TEXT, inputField!!.text.toString())
    }

    internal override fun restoreState(savedState: Bundle) {
        super.restoreState(savedState)
        if (savedState.getBoolean(KEY_HAS_ERROR, false)) {
            setError()
        }
        inputField!!.setText(savedState.getString(KEY_TYPED_TEXT))
    }

    private inner class TextInputListener (private val wrapped: OnTextInputConfirmListener?) : View.OnClickListener {

        override fun onClick(v: View) {
            val text = inputField!!.text.toString()

            if (filter != null) {
                val isWrongInput = !filter!!.check(text)
                if (isWrongInput) {
                    setError()
                    return
                }
            }

            wrapped?.onTextInputConfirmed(text)

            dismiss()
        }
    }

    private inner class HideErrorOnTextChanged : TextWatcher {

        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

        }

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            hideError()
        }

        override fun afterTextChanged(s: Editable) {

        }
    }

    interface OnTextInputConfirmListener {
        fun onTextInputConfirmed(text: String)
    }

    interface TextFilter {
        fun check(text: String): Boolean
    }

    companion object {

        private val KEY_HAS_ERROR = "key_has_error"
        private val KEY_TYPED_TEXT = "key_typed_text"
    }
}
